File: java/org.financer.client/src/main/java/org/financer/client/domain/model/category/Category.java
package org.financer.client.domain.model.category;

import org.financer.client.domain.model.transaction.Transaction;
import org.financer.client.domain.model.user.User;
import org.financer.shared.domain.model.AmountProvider;
import org.financer.shared.domain.model.Formattable;
import org.financer.shared.domain.model.Settings;
import org.financer.shared.domain.model.value.objects.Amount;
import org.financer.shared.domain.model.value.objects.CategoryClass;
import org.financer.shared.domain.model.value.objects.TimeRange;
import org.financer.shared.domain.model.value.objects.ValueDate;
import org.financer.util.collections.Tree;

import java.io.Serializable;
import java.util.*;

public class Category implements Serializable, Tree, AmountProvider, Formattable {
    private static final long serialVersionUID = 5491420625985358596L;

    private long id;
    private User user;
    private CategoryClass categoryClass;
    private Category parent;
    private String name;
    private Set<Category> children;
    private Set<Transaction> transactions = new HashSet<>();

    public String getPrefix() {
        if (this.isRoot()) {
            return Integer.toString(categoryClass.getCategoryClass().ordinal() + 1);
        }

        List<Category> neighbors = new ArrayList<>(parent.getChildren());
        neighbors.sort(Comparator.comparing(Category::getName));
        final int index = neighbors.indexOf(this) + 1;

        return parent.getPrefix() + "." + index;
    }

    @Override
    public Amount getAmount() {
        Amount amount = new Amount();

        for (AmountProvider amountProvider : this.transactions) {
            amount = amount.add(amountProvider.getAmount());
        }
        if (this.isLeaf()) {
            for (AmountProvider amountProvider : this.children) {
                amount = amount.add(amountProvider.getAmount());
            }
        }

        return amount;
    }

    @Override
    public Amount getAmount(ValueDate valueDate) {
        Amount amount = new Amount();

        for (AmountProvider amountProvider : this.transactions) {
            amount = amount.add(amountProvider.getAmount(valueDate));
        }
        if (this.isLeaf()) {
            for (AmountProvider amountProvider : this.children) {
                amount = amount.add(amountProvider.getAmount(valueDate));
            }
        }

        return amount;
    }

    @Override
    public Amount getAmount(TimeRange timeRange) {
        Amount amount = new Amount();

        for (AmountProvider amountProvider : this.transactions) {
            amount = amount.add(amountProvider.getAmount(timeRange));
        }
        if (!this.isLeaf()) {
            for (AmountProvider amountProvider : this.children) {
                amount = amount.add(amountProvider.getAmount(timeRange));
            }
        }

        return amount;
    }

    @Override
    public boolean isFixed() {
        return this.categoryClass.isFixed();
    }

    @Override
    public boolean isRevenue() {
        return this.categoryClass.isRevenue();
    }

    @Override
    public void adjustAmountSign() {
        for (AmountProvider amountProvider : this.transactions) {
            amountProvider.adjustAmountSign();
        }
        if (!this.isLeaf()) {
            for (AmountProvider amountProvider : this.children) {
                amountProvider.adjustAmountSign();
            }
        }
    }

    @Override
    public String format(Settings settings) {
        final String prefix = this.getPrefix();
        if (!prefix.isBlank()) {
            return this.getPrefix() + " " + this.getName();
        }
        return this.name;
    }

    /*
     * Getters and Setters
     */

    public long getId() {
        return id;
    }

    public Category setId(long id) {
        this.id = id;
        return this;
    }

    public User getUser() {
        return user;
    }

    public Category setUser(User user) {
        this.user = user;
        return this;
    }

    public CategoryClass getCategoryClass() {
        return categoryClass;
    }

    public Category setCategoryClass(CategoryClass categoryClass) {
        this.categoryClass = categoryClass;
        return this;
    }

    @Override
    public Category getParent() {
        return parent;
    }

    @Override
    public Category setParent(Tree parent) {
        this.parent = (Category) parent;
        return this;
    }


    public String getName() {
        return name;
    }

    public Category setName(String name) {
        this.name = name;
        return this;
    }

    @Override
    public Set<Category> getChildren() {
        return children;
    }

    public Category setChildren(Set<Category> children) {
        this.children = children;
        return this;
    }

    public Set<Transaction> getTransactions() {
        return transactions;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Category category = (Category) o;
        return id == category.id;
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    @Override
    public String toString() {
        return "Category[" + "id=" + id + ", categoryClass=" + categoryClass + ", name='" + name + '\'' + ']';
    }
}

File: java/org.financer.client/src/main/kotlin/org/financer/client/domain/model/category/Category.kt
package org.financer.client.domain.model.category

import org.financer.client.domain.model.transaction.Transaction
import org.financer.client.domain.model.user.User
import org.financer.shared.domain.model.AmountProvider
import org.financer.shared.domain.model.Formattable
import org.financer.shared.domain.model.Settings
import org.financer.shared.domain.model.value.objects.Amount
import org.financer.shared.domain.model.value.objects.CategoryClass
import org.financer.shared.domain.model.value.objects.TimeRange
import org.financer.shared.domain.model.value.objects.ValueDate
import org.financer.util.collections.Tree
import java.io.Serializable
import java.util.*

class Category : Serializable, Tree, AmountProvider, Formattable {
    /*
     * Getters and Setters
     */ var id: Long = 0
        private set
    var user: User? = null
        private set
    var categoryClass: CategoryClass? = null
        private set
    private var parent: Category? = null
    var name: String? = null
        private set
    private var children: Set<Category>? = null
    val transactions: MutableSet<Transaction> = HashSet()
    val prefix: String
        get() {
            if (this.isRoot) {
                return (categoryClass!!.categoryClass.ordinal + 1).toString()
            }
            val neighbors: List<Category> = ArrayList(parent!!.getChildren())
            neighbors.sortedWith(compareBy { it.name })
            val index = neighbors.indexOf(this) + 1
            return parent!!.prefix + "." + index
        }

    override fun getAmount(): Amount {
        var amount = Amount()
        for (amountProvider in transactions) {
            amount = amount.add(amountProvider.amount)
        }
        if (!this.isLeaf) {
            for (amountProvider in children!!) {
                amount = amount.add(amountProvider.amount)
            }
        }
        return amount
    }

    override fun getAmount(valueDate: ValueDate): Amount {
        var amount = Amount()
        for (amountProvider in transactions) {
            amount = amount.add(amountProvider.getAmount(valueDate))
        }
        if (this.isLeaf) {
            for (amountProvider in children!!) {
                amount = amount.add(amountProvider.getAmount(valueDate))
            }
        }
        return amount
    }

    override fun getAmount(timeRange: TimeRange): Amount {
        var amount = Amount()
        for (amountProvider in transactions) {
            amount = amount.add(amountProvider.getAmount(timeRange))
        }
        if (!this.isLeaf) {
            for (amountProvider in children!!) {
                amount = amount.add(amountProvider.getAmount(timeRange))
            }
        }
        return amount
    }

    override fun isFixed(): Boolean {
        return categoryClass!!.isFixed
    }

    override fun isRevenue(): Boolean {
        return categoryClass!!.isRevenue
    }

    override fun adjustAmountSign() {
        for (amountProvider in transactions) {
            amountProvider.adjustAmountSign()
        }
        if (!this.isLeaf) {
            for (amountProvider in children!!) {
                amountProvider.adjustAmountSign()
            }
        }
    }

    override fun format(settings: Settings): String {
        val prefix = prefix
        return if (!prefix.isBlank()) {
            this.prefix + " " + name
        } else name!!
    }

    fun setId(id: Long): Category {
        this.id = id
        return this
    }

    fun setUser(user: User?): Category {
        this.user = user
        return this
    }

    fun setCategoryClass(categoryClass: CategoryClass?): Category {
        this.categoryClass = categoryClass
        return this
    }

    override fun getParent(): Category {
        return parent!!
    }

    override fun setParent(parent: Tree): Category {
        this.parent = parent as Category
        return this
    }

    fun setName(name: String?): Category {
        this.name = name
        return this
    }

    override fun getChildren(): Set<Category> {
        return children!!
    }

    fun setChildren(children: Set<Category>?): Category {
        this.children = children
        return this
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val category = o as Category
        return id == category.id
    }

    override fun hashCode(): Int {
        return Objects.hash(id)
    }

    override fun toString(): String {
        return "Category[id=$id, categoryClass=$categoryClass, name='$name']"
    }

    companion object {
        private const val serialVersionUID = 5491420625985358596L
    }
}